### PR TITLE
[SPARK-19347] ReceiverSupervisorImpl can add block to ReceiverTracker multiple times because of askWithRetry.

### PR DESCRIPTION
**Pull request by jinxing64**
## What changes were proposed in this pull request?

`ReceiverSupervisorImpl` on executor side reports block's meta back to `ReceiverTracker` on driver side. In current code, `askWithRetry` is used. However, for `AddBlock`, `ReceiverTracker` is not idempotent, which may result in messages are processed multiple times.

**To reproduce**:

1. Check if it is the first time receiving `AddBlock` in `ReceiverTracker`, if so sleep long enough(say 200 seconds), thus the first RPC call will be timeout in `askWithRetry`, then `AddBlock` will be resent.
2. Rebuild Spark and run following job:
```
  def streamProcessing(): Unit = {
    val conf = new SparkConf()
      .setAppName("StreamingTest")
      .setMaster(masterUrl)
    val ssc = new StreamingContext(conf, Seconds(200))
    val stream = ssc.socketTextStream("localhost", 1234)
    stream.print()
    ssc.start()
    ssc.awaitTermination()
  }
```
**To fix**:

It makes sense to provide a blocking version `ask` in RpcEndpointRef, as mentioned in SPARK-18113 (https://github.com/apache/spark/pull/16503#event-927953218). Because Netty RPC layer will not drop messages. `askWithRetry` is a leftover from akka days. It imposes restrictions on the caller(e.g. idempotency) and other things that people generally don't pay that much attention to when using it.

## How was this patch tested?
Test manually. The scenario described above doesn't happen with this patch.

### CODE REVIEW

**Comment by squito** [comment]
Jenkins, ok to test

**Comment by squito** [comment]
cc @vanzin @zsxwing 

(I was going to make a longer comment about removing `askWithRetry` and whether we need another method, but then saw the comments on the referenced PR -- I'll defer to the experts here)

**Comment by jinxing64** [comment]
@vanzin @zsxwing 
ping for review~

File: core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala [hunk 1]
```diff
@@ -19,11 +19,13 @@ package org.apache.spark.rpc
 
 import scala.concurrent.Future
 import scala.reflect.ClassTag
+import scala.util.control.NonFatal
 
 import org.apache.spark.{SparkConf, SparkException}
 import org.apache.spark.internal.Logging
 import org.apache.spark.util.RpcUtils
 
+
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala]: nit: don't add.

File: core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala [hunk 2]
```diff
@@ -63,8 +65,48 @@ private[spark] abstract class RpcEndpointRef(conf: SparkConf)
...
+   * default timeout, throw a SparkException if this fails.
+   *
+   * Note: this is a blocking action which may cost a lot of time,  so don't call it in a message
+   * loop of [[RpcEndpoint]].
+
+   * @param message the message to send
+   * @tparam T type of the reply message
+   * @return the reply message from the corresponding [[RpcEndpoint]]
+   */
+  def askWithBlocking[T: ClassTag](message: Any): T = askWithBlocking(message, defaultAskTimeout)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala]: `askWithBlocking` is a weird name. I'd use `blockingAsk`, or `askSync`.

File: core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala [hunk 3]
```diff
@@ -63,8 +65,48 @@ private[spark] abstract class RpcEndpointRef(conf: SparkConf)
...
+   * @param message the message to send
+   * @param timeout the timeout duration
+   * @tparam T type of the reply message
+   * @return the reply message from the corresponding [[RpcEndpoint]]
+   */
+  def askWithBlocking[T: ClassTag](message: Any, timeout: RpcTimeout): T = {
+    try {
+      val future = ask[T](message, timeout)
+      val result = timeout.awaitResult(future)
+      if (result == null) {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala]: This is not an error. It's perfectly legitimate to return null.

File: core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala [hunk 4]
```diff
@@ -63,8 +65,48 @@ private[spark] abstract class RpcEndpointRef(conf: SparkConf)
...
+        throw new SparkException("RpcEndpoint returned null")
+      }
+      return result
+    } catch {
+      case NonFatal(e) =>
+        throw new SparkException(
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala]: Isn't it better to just propagate the original exception? You can get the context from the stack trace.

**Review by vanzin — CHANGES_REQUESTED** [review]
Also, since you're adding the blocking API, please add a deprecation annotation to `askWithRetry`.

**Comment by jinxing64** [comment]
@vanzin 
ping for review

**Comment by vanzin** [comment]
> ping for review

Please be a little more patient, especially during weekends.

**Comment by jinxing64** [comment]
I feel very sorry if this is disturbing : )
@vanzin Thanks a lot for continuing reviewing this and I'll be more patient : )
Sorry again~~

File: core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala [hunk 5]
```diff
@@ -91,6 +123,7 @@ private[spark] abstract class RpcEndpointRef(conf: SparkConf)
    * @tparam T type of the reply message
    * @return the reply message from the corresponding [[RpcEndpoint]]
    */
+  @deprecated("use 'askSync' instead.", "2.1.0")
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala]: 2.2.0

File: core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala [hunk 6]
```diff
@@ -75,10 +106,11 @@ private[spark] abstract class RpcEndpointRef(conf: SparkConf)
...
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala]: 2.2.0

File: core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala [hunk 7]
```diff
@@ -19,6 +19,7 @@ package org.apache.spark.rpc
 
 import scala.concurrent.Future
 import scala.reflect.ClassTag
+import scala.util.control.NonFatal
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala]: not used

**Review by vanzin — APPROVED** [review]
Looks good, just need to fix the versions.

**Comment by jinxing64** [comment]
@vanzin 
Thanks a lot for helping this PR~ I've already refined~
Please take another look~

**Comment by vanzin** [comment]
LGTM, merging to master.

**Comment by jinxing64** [comment]
Thanks a lot for reviewing this PR~

File: core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala [hunk 8]
```diff
@@ -75,10 +105,11 @@ private[spark] abstract class RpcEndpointRef(conf: SparkConf)
    * @tparam T type of the reply message
    * @return the reply message from the corresponding [[RpcEndpoint]]
    */
+  @deprecated("use 'askSync' instead.", "2.2.0")
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala]: It seems like this has caused the build to produce a lot of deprecation warnings. @jinxing64 could the callers of this method be changed, in Spark, to use the new alternative?

**Comment by jinxing64** [comment]
@srowen 
How do you think about https://github.com/apache/spark/pull/16790?

**Pull request closed without merging**